### PR TITLE
[GHA] Terminate ci-test workflow earlier once there is job failed

### PR DESCRIPTION
**Pull request by zihaoccc**
Instead always run full workflow each time which takes around 50mins, we can terminate it once there have any job is failed.

https://github.com/diem/diem/actions/runs/1401703623

## Motivation

(Write your motivation for proposed changes here.)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/diem/diem/tree/main/developers.diem.com, and link to your PR here.)

## If targeting a release branch, please fill the below out as well

 * Justification and breaking nature (who does it affect? validators, full nodes, tooling, operators, AOS, etc.)
 * Comprehensive test results that demonstrate the fix working and not breaking existing workflows.
 * Why we must have it for V1 launch.
 * What workarounds and alternative we have if we do not push the PR.

### CODE REVIEW

File: testsuite/cluster-test/src/experiments/reconfiguration_test.rs [hunk 1]
```diff
@@ -19,7 +19,7 @@ use diem_types::{
     account_address::AccountAddress,
     chain_id::ChainId,
     ledger_info::LedgerInfoWithSignatures,
-    on_chain_config::{ConsensusConfigV2, OnChainConsensusConfig},
+    on_chain_config::{ConsensusConfigV1, OnChainConsensusConfig},
```
Comment (bmwill) [inline testsuite/cluster-test/src/experiments/reconfiguration_test.rs:22]: The code in this file seems unrelated to the PR?

Comment (zihaoccc) [reply testsuite/cluster-test/src/experiments/reconfiguration_test.rs:22]: ops, maybe rebase issue

Comment (zihaoccc) [reply testsuite/cluster-test/src/experiments/reconfiguration_test.rs]: fixed

File: .github/actions/early-terminator/action.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  steps:
+    - name: Post test results on PR
+      uses: actions/github-script@v4.0.2
+      with:
+        script: |
+            // use github REST apis: https://docs.github.com/en/rest/reference/actions#workflow-runs
+            const https = require('https');
+            const options = {
+              hostname: 'api.github.com',
+              path: `/repos/${process.env.GITHUB_REPOSITORY}/actions/runs/${process.env.GITHUB_RUN_ID}/cancel`,
```
Comment (sausagee) [inline .github/actions/early-terminator/action.yml:19]: We send the signal to cancel the workflow, but would Bors continue to listen for it?  Put another way, do we need a hook to tell Bors?

Comment (bmwill) [reply .github/actions/early-terminator/action.yml:19]: Nope, that's actually the benefit of this approach. If you look at the above tests Zihao did there's a ~10min difference between when bors started the canary and when bors posted that there was a failure (instead of the usual 50+). The biggest question I have is, what happens to LBT when canceled? Is the cluster freed immediately?

File: .github/actions/early-terminator/action.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+              headers: {
+                'Authorization': `token ${{ inputs.github-token }}`,
```
Comment (bmwill) [inline .github/actions/early-terminator/action.yml:21]: Given this uses the github token, do we know how this is supposed to behave on Ci run on PRs?

File: .github/actions/early-terminator/action.yml (hunk 2 above)
Comment (zihaoccc) [reply .github/actions/early-terminator/action.yml:19]: Bors will report failed as above canary cases

Comment (zihaoccc) [reply .github/actions/early-terminator/action.yml:19]: 💔 Test Failed - ci-test

File: .github/actions/early-terminator/action.yml (hunk 3 above)
Comment (zihaoccc) [reply .github/actions/early-terminator/action.yml:21]: I have done canary run with passing in GH token, https://github.com/diem/diem/actions/runs/1401703623

File: .github/actions/early-terminator/action.yml (hunk 2 above)
Comment (zihaoccc) [reply .github/actions/early-terminator/action.yml:19]: we can treat this approach as the same by clicking the cancel bottom manually, which is using the same rest api to trigger cancellation (I grabbed this api by web source code). Bors will report failed once any job is failed, so the duration is depend on when will be the first job fail, maybe from very start by few secs failed in prepare step or few 10 mins in LBT. @bmwill I think we do not really care about LBT, since LBT is the long tail, most of cases of early termination, LBT even does not be triggered yet. Or if we cancel workflow at the half of LBT runs, Forge cluster will do cleanup as well since cleanup is in destructor. Or for worst case, current LBT does not do cleanup, but next run will do cleanup as well prior the test.

Comment (zihaoccc) [reply .github/actions/early-terminator/action.yml:19]: @bmwill oh, I think I got your point, your concern is even workflow is cancelled, but cluster still keep running test until finish. I need to double check this case, this is a really good point. But still, this early termination can still gain benefits on cases which have job failed at lint, building-image, unit-test etc

File: .github/actions/early-terminator/action.yml (hunk 3 above)
Comment (bmwill) [reply .github/actions/early-terminator/action.yml:21]: No i'm specifically asking about what happens when Ci runs on a PR, not being run as apart of a land or canary step on `auto` or `canary` branch

Comment (zihaoccc) [reply .github/actions/early-terminator/action.yml:21]: this early termination is only added in ci-workflow which I assume it is only triggered by auto and canary branches? If i misunderstood please correct me

Comment (bmwill) [reply .github/actions/early-terminator/action.yml:21]: No ci-test workflow is triggered on PRs as well as on auto and canary branches

Comment (zihaoccc) [reply .github/actions/early-terminator/action.yml:21]: Since we have other actions are also using GHA token as passing in value, I think it should be ok but we can do some tests to confirm. 
```
id: changes
name: determine changes
uses: diem/actions/changes@faadd16607b77dfa2231a8f366883e01717b3225
with:
    workflow-file: ci-test.yml
    github-token: ${{secrets.GITHUB_TOKEN}}

LOGIN=( "--header" "authorization: Bearer ${GITHUB_TOKEN}" )
```

**Review by bmwill — APPROVED** [review]
/land p=high
